### PR TITLE
OperatorHub Release CLI / Moving OperatorHub release to Buildkite

### PR DESCRIPTION
**Pull request by naemono**
This adds a Buildkite pipeline for release operations pertaining to OperatorHub/Red Hat.
This also augments our existing OpertaorHub CLI and adds new commands/sub-commands.

~~What's been manually tested:~~

~~Full set of operatorhub release operations with a `BC` build of Operator.~~

~~In progress:~~

- [x] ~~Currently testing Buildkite integration (will fully test during actual release).~~

^^ All fully tested during release process for 2.6.x.

### CODE REVIEW

File: .buildkite/pipeline.yml
Comment (naemono) [inline .buildkite/pipeline.yml:135]: I wonder here if pulling the preflight binary while running with our default agent image (or adding the preflight binary to our agent) would be a better alternative than what's being done below.  Would love thoughts about this approach.

File: hack/operatorhub/internal/operatorhub/operatorhub.go
Comment (thbkrkr) [inline hack/operatorhub/internal/operatorhub/operatorhub.go:273]: The fix to ensure local YAML files have a proper "end of directives" marker (#5447) is missing:
```suggestion
		rs = append(rs, r)
		// if we're using local yaml files, ensure that they have a proper
		// end of directives marker between them.
		rs = append(rs, strings.NewReader(yamlSeparator))
```

File: .buildkite/pipeline.yml
Comment (pebrc) [inline .buildkite/pipeline.yml:121]: Would it make sense to use a custom image for this to not have to build the tooling every time?

File: .buildkite/pipeline.yml
Comment (pebrc) [inline .buildkite/pipeline.yml:150]: A custom image could also contain `jq`

Comment (naemono) [reply .buildkite/pipeline.yml:150]: Agreed.  I've opened a PR to add jq to our custom image we already use @pebrc

File: .buildkite/pipeline.yml
Comment (naemono) [reply .buildkite/pipeline.yml:121]: @pebrc the build process takes ~60 seconds in BK.  If we add this binary to the image we're already using, then the process to fix any new Red Hat changes in the release process (which are pretty common) would involve:

1. opening a PR in this repository, waiting for approval/merge of the tooling.
2. opening a PR in the image repository to bump the tooling version, waiting for approval/merge.
3. Re-running the bk job.

As opposed to:
1. opening a PR in this repository with the tooling changes
2. running the bk job, pointing to the open PR for the tooling changes.  (note the `--buildkite-pr-repository` and `--buildkite-pr-id` flags in the `operatorhub buildkite` command which exist for this scenario)

File: hack/operatorhub/internal/operatorhub/operatorhub.go
Comment (thbkrkr) [inline hack/operatorhub/internal/operatorhub/operatorhub.go:319]: https://github.com/elastic/cloud-on-k8s/commit/1168b13c6606034d1d1f43fcce207918df2d2ce0

```suggestion
	defer resp.Body.Close()
```

File: hack/operatorhub/README.md
Comment (thbkrkr) [inline hack/operatorhub/README.md:127]: Why asked `-s '8.6.0' -V '2.5.0' -t '2.6.0'`when it's already defined in `-c config.yaml`?

File: .buildkite/pipeline.yml
Comment (thbkrkr) [inline .buildkite/pipeline.yml:160]: When there are so many commands, I think it will be more appropriate to write a bash script (`.buildkite/scripts/release/operatorhub-preflight.sh`). This would avoid to think about the $$ for example.

Did you consider to move this in the tool? I feel like `/usr/local/bin/operatorhub container preflight` would have been nice. I'm good with this version too.

File: .buildkite/pipeline.yml
Comment (thbkrkr) [inline .buildkite/pipeline.yml:114]: I would move the steps in a new pipeline `.buildkite/pipeline-release-operatorhub.yml` that we invoke from the root pipeline `.buildkite/pipeline.yml`.

```suggestion

      - label: "redhat operatorhub"
        if: build.message == "run operatorhub release"
        commands:
            - buildkite-agent pipeline upload .buildkite/pipeline-release-operatorhub.yml
```

File: hack/operatorhub/internal/container/container.go
Comment (thbkrkr) [inline hack/operatorhub/internal/container/container.go:92]:
```suggestion
	}

	if err = pushImageToRegistry(c); err != nil {
		log.Println("x")	
		return fmt.Errorf("failed to push image: %w", err)
	}
	log.Println("✓")	
```

File: hack/operatorhub/README.md
Comment (naemono) [reply hack/operatorhub/README.md:127]: yeah, I will get this simplified.  good point.  Unnecessary.

File: .buildkite/pipeline.yml
Comment (naemono) [reply .buildkite/pipeline.yml:160]: I'll get this contained within a script.  👍 

> Did you consider to move this in the tool? I feel like `/usr/local/bin/operatorhub container preflight` would have been nice. I'm good with this version too.

Oh yes I tried to get this fully into this tool, see https://github.com/redhat-openshift-ecosystem/openshift-preflight/issues/845

File: .buildkite/pipeline-release-operatorhub.yml
Comment (thbkrkr) [inline .buildkite/pipeline-release-operatorhub.yml:15]: Since `pipeline-release-operatorhub.yml` is uploaded only `if: build.message == "run operatorhub release"` dans `pipeline.yml`, it seems redundant to repeat the conditional for all steps.

File: hack/operatorhub/README.md
Comment (thbkrkr) [inline hack/operatorhub/README.md:130]: What is the use case that would require overriding the values defined in `config.yaml`?

File: .buildkite/scripts/operatorhub/operations.sh
Comment (thbkrkr) [inline .buildkite/scripts/operatorhub/operations.sh:54]: I would have kept these release commands in the pipeline and just use this script for the preflight commands.

File: hack/operatorhub/cmd/buildkite/buildkite.go
Comment (thbkrkr) [inline hack/operatorhub/cmd/buildkite/buildkite.go:31]: I think
```bash
curl -XPOST "https://api.buildkite.com/v2/organizations/elastic/pipelines/cloud-on-k8s-operator-release-operatorhub/builds" \
-H "Authorization: Bearer $BK_TOKEN" \
-d '{
    "commit": "2.6.1",
    "branch": "2.6"
}'
```
could have been enough.

**Review by thbkrkr — COMMENTED** [review]
Maybe I miss something but as discussed in https://github.com/elastic/cloud-on-k8s/pull/6281#discussion_r1080575799, I think we shouldn't duplicate what is defined in `./config.yaml` (`OHUB_CONF`). This configuration file is the source of truth that lives in the codebase and is updated only through pull requests.
This means that all the following flags / env vars should disappear:
* OHUB_SUPPORTED_OPENSHIFT_VERSIONS
* OHUB_STACK_VERSION
* OHUB_PREV_VERSION
* OHUB_TAG
* OHUB_BRANCH

It seems to me that having to adjust them is painful and above all prone to error.

File: hack/operatorhub/templates/package.tpl
Comment (thbkrkr) [inline hack/operatorhub/templates/package.tpl:2]: I think we need to keep the `v`.
```suggestion
- currentCSV: {{ .PackageName }}.v{{ .NewVersion }}
```

Otherwise, I get:

```
> OHUB_TAG=2.6.1 hack/operatorhub/bin/operatorhub bundle generate --dir=.
2023/02/10 15:19:38 Generating operator bundle image for publishing
2023/02/10 15:19:38 ⅹ
Error: during opm generate: error interpreting channels, please manually input channels instead
2023/02/10 15:19:38 failed to run command: during opm generate: error interpreting channels, please manually input channels instead
```

Because we use the `v`prefix to name the CSV:
```
# templates/csv.tpl
apiVersion: operators.coreos.com/v1alpha1
kind: ClusterServiceVersion
metadata:
  name: {{ .PackageName }}.v{{ .NewVersion }}
```

File: .buildkite/pipeline-release-operatorhub.yml
Comment (naemono) [reply .buildkite/pipeline-release-operatorhub.yml:19]: I swore I tried this, and it wasn't properly working, but I'll certainly try again, as it makes total sense.

File: .buildkite/scripts/operatorhub/operations.sh
Comment (naemono) [reply .buildkite/scripts/operatorhub/operations.sh:1]: License header also added to `.buildkite/scripts/common/get-test-artifacts.sh`.

File: hack/operatorhub/README.md
Comment (naemono) [reply hack/operatorhub/README.md]: I think I was wanting to ensure that the settings were absolutely correct when running in an automated fashion, but since the typical use case for this is to run this from a branch that already has `config.yaml` updated appropriately, then we can likely do without these, agreed.   They've been removed.

File: .buildkite/scripts/operatorhub/operations.sh
Comment (naemono) [reply .buildkite/scripts/operatorhub/operations.sh:46]: I figured since I was cleaning up the pipeline, I might as well keep it as clean as possible, hence moving all of this to a script.  Lmk if you need me to move it back.

File: hack/operatorhub/cmd/buildkite/buildkite.go
Comment (naemono) [reply hack/operatorhub/cmd/buildkite/buildkite.go:31]: We discussed this privately, and decided that opening #6408 would be beneficial.

**Comment by naemono** [comment]
> Maybe I miss something but as discussed in [#6281 (comment)](https://github.com/elastic/cloud-on-k8s/pull/6281#discussion_r1080575799), I think we shouldn't duplicate what is defined in `./config.yaml` (`OHUB_CONF`). This configuration file is the source of truth that lives in the codebase and is updated only through pull requests. This means that all the following flags / env vars should disappear:
> 
> * OHUB_SUPPORTED_OPENSHIFT_VERSIONS
> * OHUB_STACK_VERSION
> * OHUB_PREV_VERSION
> * OHUB_TAG
> * OHUB_BRANCH
> 
> It seems to me that having to adjust them is painful and above all prone to error.

@thbkrkr I've removed prev/stack, but removing the others is a much bigger change as, for example, tag is used throughout the tool.  I'll get these removed/updated soon.

**Comment by thbkrkr** [comment]
> Unfortunately OHUB_SUPPORTED_OPENSHIFT_VERSIONS is required, as there's no setting for it in the configuration file. If you want, I can certainly add that as a new option there and remove the cli flag. LMK.

I think it's good to have all the configuration in the source files, so I feel like saying yes please.
Should it be more something like `minSupportedOpenShiftVersion` as we don't want the range here?

We already have the range in the template that we could half-generate from it or accept to have the info duplicated (we already have the minKubeVersion duplicated in the template).
https://github.com/elastic/cloud-on-k8s/blob/d8d38f5130725c2872c7aba6241ead5dc6ffba94/hack/operatorhub/templates/csv.tpl#L270

**Comment by naemono** [comment]
> We already have the range in the template that we could half-generate from it or accept to have the info duplicated (we already have the minKubeVersion duplicated in the template).

I've kept it simple, and just added it to the configuration file, and updated the validation to ensure that it exists, and is not a range.

**Review by thbkrkr — APPROVED** [review]
LGTM!


The vault bits could benefit from #6418 but we shouldn't block for that.

**Pull request merged**
